File: push_swap_prototype/internal/solver/chunk_sort.go
package solver

import (
	"fmt"
	"push_swap_prototype/internal/stack"
)

// chunk_sort sorts the entire stack A using chunk-based sorting
func chunk_sort(ps *stack.PushSwapData) {
	fmt.Printf("=== CHUNK_SORT START ===\n")
	fmt.Printf("Initial stack A size: %d\n", ps.A.Size())
	
	chunkAll := stack.Chunk{
		Loc:  stack.TOP_A, 
		Size: ps.A.Size(),
	}
	
	fmt.Printf("Initial chunk: loc=%v, size=%d\n", chunkAll.Loc, chunkAll.Size)
	recChunkSort(ps, &chunkAll)
	fmt.Printf("=== CHUNK_SORT END ===\n")
}

// recChunkSort recursively sorts chunks using divide and conquer approach
func recChunkSort(ps *stack.PushSwapData, toSort *stack.Chunk) {
	fmt.Printf("=== REC_CHUNK_SORT ===\n")
	fmt.Printf("Input chunk: loc=%v, size=%d\n", toSort.Loc, toSort.Size)
	fmt.Printf("Stack A size: %d, Stack B size: %d\n", ps.A.Size(), ps.B.Size())
	
	// Move chunk to the top if needed
	chunkToTheTop(ps, toSort)
	
	// Apply easy sort optimization
	easySort(ps, toSort)
	
	// Check if chunk size was reduced to <= 3 by easySort
	if toSort.Size <= 3 {
		fmt.Printf("Chunk size reduced to %d by easySort, handling directly\n", toSort.Size)
		// Handle small chunks with direct sorting
		switch toSort.Size {
		case 3:
			fmt.Printf("Sorting 3 elements\n")
			sortThree(ps, toSort)
		case 2:
			fmt.Printf("Sorting 2 elements\n")
			sortTwo(ps, toSort)
		case 1:
			fmt.Printf("Sorting 1 element\n")
			sortOne(ps, toSort)
		case 0:
			fmt.Printf("Chunk is empty, nothing to sort\n")
		}
		fmt.Printf("=== REC_CHUNK_SORT END (small chunk) ===\n")
		return
	}
	
	// Split chunk and recursively sort
	fmt.Printf("Splitting chunk (size > 3)\n")
	
	var dest stack.SplitDest
	chunkSplit(ps, toSort, &dest)
	
	fmt.Printf("After split - Max: %d, Mid: %d, Min: %d\n", 
		dest.Max.Size, dest.Mid.Size, dest.Min.Size)
	
	// Recursively sort the three parts (exactly like C implementation)
	fmt.Printf("Recursively sorting MAX chunk\n")
	recChunkSort(ps, &dest.Max)
	fmt.Printf("Recursively sorting MID chunk\n")
	recChunkSort(ps, &dest.Mid)
	fmt.Printf("Recursively sorting MIN chunk\n")
	recChunkSort(ps, &dest.Min)
	
	fmt.Printf("=== REC_CHUNK_SORT END ===\n")
}

// sortTwo sorts two elements in a chunk
func sortTwo(ps *stack.PushSwapData, chunk *stack.Chunk) {
	fmt.Printf("  sortTwo: chunk loc=%v, size=%d\n", chunk.Loc, chunk.Size)
	
	// Move elements to TOP_A if they're not already there (exactly like C implementation)
	if chunk.Loc == stack.BOTTOM_A || chunk.Loc == stack.BOTTOM_B || chunk.Loc == stack.TOP_B {
		fmt.Printf("    Moving elements to TOP_A\n")
		moveFromTo(ps, chunk.Loc, stack.TOP_A)
		moveFromTo(ps, chunk.Loc, stack.TOP_A)
	}
	
	// If the top element is greater than the second element, swap them
	// Use 1-based indexing like C implementation
	if ps.A.GetValueAtPosition(1) > ps.A.GetValueAtPosition(2) {
		fmt.Printf("    Swapping A[0] and A[1]\n")
		stack.Swap_a(ps)
	}
	
	chunk.Size -= 2
	fmt.Printf("    After sortTwo: chunk size=%d\n", chunk.Size)
}

// sortOne handles a single element in a chunk
func sortOne(ps *stack.PushSwapData, chunk *stack.Chunk) {
	fmt.Printf("  sortOne: chunk loc=%v, size=%d\n", chunk.Loc, chunk.Size)
	
	// Move element to TOP_A if it's not already there
	if chunk.Loc == stack.BOTTOM_A || chunk.Loc == stack.BOTTOM_B || chunk.Loc == stack.TOP_B {
		fmt.Printf("    Moving element to TOP_A\n")
		moveFromTo(ps, chunk.Loc, stack.TOP_A)
	}
	chunk.Size--
	fmt.Printf("    After sortOne: chunk size=%d\n", chunk.Size)
}

File: push_swap_prototype/lab/src/push_swap/chunk_sort.c
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   chunk_sort.c                                       :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: patrik <patrik@student.42.fr>              +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2023/07/27 20:56:55 by ugerkens          #+#    #+#             */
/*   Updated: 2025/08/25 11:01:26 by patrik           ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include "../../include/push_swap.h"
#include "../../include/debug_logging.h"

void	chunk_sort(t_ps *data)
{
	t_chunk	chunk_all;

	chunk_all.loc = TOP_A;
	chunk_all.size = data->a.size;
	rec_chunk_sort(data, &chunk_all);
}

void	rec_chunk_sort(t_ps *data, t_chunk *to_sort)
{
	t_split_dest	dest;

	// Debug: Log function entry
	debug_log(DEBUG_DEBUG, "=== REC_CHUNK_SORT START ===");
	debug_log_chunk_state(to_sort, "to_sort at entry");
	debug_log(DEBUG_DEBUG, "to_sort->size = %d", to_sort->size);

	chunk_to_the_top(data, to_sort);
	debug_log(DEBUG_DEBUG, "After chunk_to_the_top: to_sort->loc=%d, to_sort->size=%d", to_sort->loc, to_sort->size);
	
	easy_sort(data, to_sort);
	debug_log(DEBUG_DEBUG, "After easy_sort: to_sort->loc=%d, to_sort->size=%d", to_sort->loc, to_sort->size);

	// Debug: Log before size check
	debug_log(DEBUG_DEBUG, "Before size check: to_sort->size = %d", to_sort->size);
	
	if (to_sort->size <= 3)
	{
		debug_log(DEBUG_DEBUG, "Size <= 3, processing directly");
		
		if (to_sort->size == 3)
		{
			debug_log(DEBUG_DEBUG, "Calling sort_three");
			sort_three(data, to_sort);
		}
		else if (to_sort->size == 2)
		{
			debug_log(DEBUG_DEBUG, "Calling sort_two");
			sort_two(data, to_sort);
		}
		else if (to_sort->size == 1)
		{
			debug_log(DEBUG_DEBUG, "Calling sort_one");
			sort_one(data, to_sort);
		}
		debug_log(DEBUG_DEBUG, "=== REC_CHUNK_SORT END (size <= 3) ===");
		return ;
	}
	
	debug_log(DEBUG_DEBUG, "Size > 3, calling chunk_split");
	chunk_split(data, to_sort, &dest);
	
	// Debug: Log destination chunks after split
	debug_log(DEBUG_DEBUG, "After chunk_split:");
	debug_log_chunk_state(&dest.max, "dest.max");
	debug_log_chunk_state(&dest.mid, "dest.mid");
	debug_log_chunk_state(&dest.min, "dest.min");
	
	debug_log(DEBUG_DEBUG, "Calling rec_chunk_sort with dest.max");
	rec_chunk_sort(data, &dest.max);
	
	debug_log(DEBUG_DEBUG, "Calling rec_chunk_sort with dest.mid");
	rec_chunk_sort(data, &dest.mid);
	
	debug_log(DEBUG_DEBUG, "Calling rec_chunk_sort with dest.min");
	rec_chunk_sort(data, &dest.min);
	
	debug_log(DEBUG_DEBUG, "=== REC_CHUNK_SORT END (recursive) ===");
}

void	sort_two(t_ps *data, t_chunk *to_sort)
{
	// Debug: Log function entry
	debug_log(DEBUG_DEBUG, "=== SORT_TWO START ===");
	debug_log_chunk_state(to_sort, "to_sort at entry");

	if (to_sort->loc == BOTTOM_A || to_sort->loc == BOTTOM_B
		|| to_sort->loc == TOP_B)
	{
		debug_log(DEBUG_DEBUG, "Case: to_sort->loc != TOP_A, calling move_from_to twice");
		debug_log(DEBUG_DEBUG, "First move_from_to: %d -> TOP_A", to_sort->loc);
		move_from_to(data, to_sort->loc, TOP_A);
		debug_log(DEBUG_DEBUG, "Second move_from_to: %d -> TOP_A", to_sort->loc);
		move_from_to(data, to_sort->loc, TOP_A);
	}
	else
	{
		debug_log(DEBUG_DEBUG, "Case: to_sort->loc == TOP_A, no move_from_to needed");
	}
	
	debug_log(DEBUG_DEBUG, "Checking if value(&data->a, 1) > value(&data->a, 2)");
	debug_log(DEBUG_DEBUG, "value(&data->a, 1) = %d, value(&data->a, 2) = %d", value(&data->a, 1), value(&data->a, 2));
	
	if (value(&data->a, 1) > value(&data->a, 2))
	{
		debug_log(DEBUG_DEBUG, "Values are not in order, calling swap_a");
		swap_a(data);
	}
	else
	{
		debug_log(DEBUG_DEBUG, "Values are already in order, no swap needed");
	}
	
	to_sort->size -= 2;
	debug_log(DEBUG_DEBUG, "Decremented to_sort->size by 2, new size = %d", to_sort->size);
	
	debug_log(DEBUG_DEBUG, "=== SORT_TWO END ===");
}

void	sort_one(t_ps *data, t_chunk *to_sort)
{
	// Debug: Log function entry
	debug_log(DEBUG_DEBUG, "=== SORT_ONE START ===");
	debug_log_chunk_state(to_sort, "to_sort at entry");
	
	if (to_sort->loc == BOTTOM_A || to_sort->loc == BOTTOM_B
		|| to_sort->loc == TOP_B)
	{
		debug_log(DEBUG_DEBUG, "Case: to_sort->loc != TOP_A, calling move_from_to");
		debug_log(DEBUG_DEBUG, "move_from_to: %d -> TOP_A", to_sort->loc);
		move_from_to(data, to_sort->loc, TOP_A);
	}
	else
	{
		debug_log(DEBUG_DEBUG, "Case: to_sort->loc == TOP_A, no move_from_to needed");
	}
	
	to_sort->size -= 1;
	debug_log(DEBUG_DEBUG, "Decremented to_sort->size by 1, new size = %d", to_sort->size);
	
	debug_log(DEBUG_DEBUG, "=== SORT_ONE END ===");
}
